File: src/kernel/primitives/capabilities.ts
// Closed capability vocabulary for the kernel domain model. This intentionally
// mirrors src/capability/capabilities.ts but is defined here, standalone, so
// the kernel primitives compile with zero dependency on any agent provider,
// planner, or runtime module (Acceptance Criterion 1). The kernel is the
// future source of truth; the runtime copy stays in sync via adapters (a
// later milestone), not via a compile-time import in this direction.

export type Capability =
  | 'coding'
  | 'reasoning'
  | 'planning'
  | 'review'
  | 'docs'
  | 'translation'
  | 'vision'
  | 'audio'
  | 'embeddings'
  | 'search'
  | 'locate'

export const ALL_CAPABILITIES: readonly Capability[] = [
  'coding', 'reasoning', 'planning', 'review', 'docs', 'translation',
  'vision', 'audio', 'embeddings', 'search', 'locate',
]

export function isCapability(v: unknown): v is Capability {
  return typeof v === 'string' && (ALL_CAPABILITIES as readonly string[]).includes(v)
}

// Typed artifact kinds a task may produce or require. Mirrors
// src/artifact/artifacts.ts ArtifactKind for the same standalone reason.
export type ArtifactKind =
  | 'triage'
  | 'grill'
  | 'context'
  | 'patch'
  | 'plan'
  | 'execution'
  | 'evaluation'
  | 'final'
  | 'decision'
  | 'review'
  | 'test-result'
  | 'pointer-set'
  | 'token-estimate'
  | 'compression'
  | 'cost'
  | 'intent'
  | 'metric'
  | 'failure'
  | 'clarification'

export const ALL_ARTIFACT_KINDS: readonly ArtifactKind[] = [
  'triage', 'grill', 'context', 'patch', 'plan', 'execution',
  'evaluation', 'final', 'decision', 'review', 'test-result',
  'pointer-set', 'token-estimate', 'compression', 'cost',
  'intent', 'metric', 'failure', 'clarification',
]

export function isArtifactKind(v: unknown): v is ArtifactKind {
  return typeof v === 'string' && (ALL_ARTIFACT_KINDS as readonly string[]).includes(v)
}

File: cortex-types/src/capabilities.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Capability {
    Coding,
    Reasoning,
    Planning,
    Review,
    Docs,
    Translation,
    Vision,
    Audio,
    Embeddings,
    Search,
    Locate,
}

pub const ALL_CAPABILITIES: &[Capability] = &[
    Capability::Coding,
    Capability::Reasoning,
    Capability::Planning,
    Capability::Review,
    Capability::Docs,
    Capability::Translation,
    Capability::Vision,
    Capability::Audio,
    Capability::Embeddings,
    Capability::Search,
    Capability::Locate,
];

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ArtifactKind {
    Triage,
    Grill,
    Context,
    Patch,
    Plan,
    Execution,
    Evaluation,
    Final,
    Decision,
    Review,
    TestResult,
    PointerSet,
    TokenEstimate,
    Compression,
    Cost,
    Intent,
    Metric,
    Failure,
    Clarification,
}

pub fn is_capability(value: &str) -> bool {
    matches!(
        value,
        "coding"
            | "reasoning"
            | "planning"
            | "review"
            | "docs"
            | "translation"
            | "vision"
            | "audio"
            | "embeddings"
            | "search"
            | "locate"
    )
}

pub const ALL_ARTIFACT_KINDS: &[ArtifactKind] = &[
    ArtifactKind::Triage,
    ArtifactKind::Grill,
    ArtifactKind::Context,
    ArtifactKind::Patch,
    ArtifactKind::Plan,
    ArtifactKind::Execution,
    ArtifactKind::Evaluation,
    ArtifactKind::Final,
    ArtifactKind::Decision,
    ArtifactKind::Review,
    ArtifactKind::TestResult,
    ArtifactKind::PointerSet,
    ArtifactKind::TokenEstimate,
    ArtifactKind::Compression,
    ArtifactKind::Cost,
    ArtifactKind::Intent,
    ArtifactKind::Metric,
    ArtifactKind::Failure,
    ArtifactKind::Clarification,
];
